From 0f1196d36df107c06a6e306c0c2bc9bbc452056a Mon Sep 17 00:00:00 2001
From: Sebastian Andrzej Siewior <bigeasy@linutronix.de>
Date: Mon, 29 Jun 2026 16:47:21 +0200
Subject: [PATCH 3/4] ARM: Fixup the stable merge.

This a merge fixup, properly replaces
	ARM: mm: fault: Move harden_branch_predictor() before interrupts are enabled
	ARM: mm: fault: Enable interrupts before invoking __do_user_fault()

with
	ARM: ensure interrupts are enabled in __do_user_fault()

It is needed because "other" patches were applied as part of the stable
series. This contains the patches that were applied upstream.

Signed-off-by: Sebastian Andrzej Siewior <bigeasy@linutronix.de>
---
 arch/arm/mm/fault.c | 19 ++++++++-----------
 1 file changed, 8 insertions(+), 11 deletions(-)

diff --git a/arch/arm/mm/fault.c b/arch/arm/mm/fault.c
index c4e01d6012fd6..1de1f3ff64eb4 100644
--- a/arch/arm/mm/fault.c
+++ b/arch/arm/mm/fault.c
@@ -190,7 +190,8 @@ __do_kernel_fault(struct mm_struct *mm, unsigned long addr, unsigned int fsr,
 
 /*
  * Something tried to access memory that isn't in our memory map..
- * User mode accesses just cause a SIGSEGV
+ * User mode accesses just cause a SIGSEGV. Ensure interrupts are enabled
+ * for preempt RT.
  */
 static void
 __do_user_fault(unsigned long addr, unsigned int fsr, unsigned int sig,
@@ -198,6 +199,8 @@ __do_user_fault(unsigned long addr, unsigned int fsr, unsigned int sig,
 {
 	struct task_struct *tsk = current;
 
+	local_irq_enable();
+
 #ifdef CONFIG_DEBUG_USER
 	if (((user_debug & UDBG_SEGV) && (sig == SIGSEGV)) ||
 	    ((user_debug & UDBG_BUS)  && (sig == SIGBUS))) {
@@ -230,14 +233,10 @@ void do_bad_area(unsigned long addr, unsigned int fsr, struct pt_regs *regs)
 	 * If we are in kernel mode at this point, we
 	 * have no context to handle this fault with.
 	 */
-	if (user_mode(regs)) {
-		if (addr >= TASK_SIZE)
-			harden_branch_predictor();
-		local_irq_enable();
+	if (user_mode(regs))
 		__do_user_fault(addr, fsr, SIGSEGV, SEGV_MAPERR, regs);
-	} else {
+	else
 		__do_kernel_fault(mm, addr, fsr, regs);
-	}
 }
 
 #ifdef CONFIG_MMU
@@ -272,6 +271,7 @@ do_kernel_address_page_fault(struct mm_struct *mm, unsigned long addr,
 		 * should not be faulting in kernel space, which includes the
 		 * vector/khelper page. Handle the branch predictor hardening
 		 * while interrupts are still disabled, then send a SIGSEGV.
+		 * Note that __do_user_fault() will enable interrupts.
 		 */
 		harden_branch_predictor();
 		__do_user_fault(addr, fsr, SIGSEGV, SEGV_MAPERR, regs);
@@ -314,11 +314,8 @@ do_page_fault(unsigned long addr, unsigned int fsr, struct pt_regs *regs)
 		return do_kernel_address_page_fault(mm, addr, fsr, regs);
 
 	/* Enable interrupts if they were enabled in the parent context. */
-	if (interrupts_enabled(regs)) {
-		if (user_mode(regs) && addr >= TASK_SIZE)
-			harden_branch_predictor();
+	if (interrupts_enabled(regs))
 		local_irq_enable();
-	}
 
 	/*
 	 * If we're in an interrupt or have no user
-- 
2.43.0

